src/common.ml

```diff
@@ -47,21 +47,28 @@ let rec string_of_exp e =
     | TermExp (f, args) ->
         let func = String.escaped f in
             "TermExp (\"" ^ func ^ "\", [" ^ (String.concat "; " (List.map string_of_exp args)) ^ "])"
-    | ConjunctionExp (e1, e2) ->
+(*    | ConjunctionExp (e1, e2) ->
         "ConjunctionExp (" ^ (string_of_exp e1) ^ ", " ^ (string_of_exp e2) ^ ")"
     | DisjunctionExp (e1, e2) ->
         "DisjunctionExp (" ^ (string_of_exp e1) ^ ", " ^ (string_of_exp e2) ^ ")"
+ *)
 
+
+let string_of_goals g = "[" ^ (String.concat "; " (List.map string_of_exp g)) ^ "]"
+                      
 let string_of_dec d =
     match d with
-    | Clause (e1, e2) -> "Clause (" ^ (string_of_exp e1) ^ ", " ^ (string_of_exp e2) ^ ")"
-    | Query e -> "Query (" ^ (string_of_exp e) ^ ")"
+    | Clause (e1, g) -> "Clause (" ^ (string_of_exp e1) ^ ", " ^ (string_of_goals g) ^ ")"
+    | Query g -> "Query (" ^ (string_of_goals g) ^ ")"
 
 let string_of_db db =
     "[" ^ (String.concat "; " (List.map string_of_dec db)) ^ "]"
 
 let print_db db =
     print_endline (string_of_db db)
 
-let string_of_res r =
-    raise (Failure "Not implemented yet")
+let string_of_atom a =
+    match a with
+    | (TermExp(s,_)) -> s
+    | _ -> raise(Failure "not needed")
```

Comment:
I would rename this `functor_of_termexp` and require the input to be a `TermExp` only.

Reply:
Restricting the type that can be passed into string_of_atom causes warnings. In fact to truly solve this issue we would need to modify our AST slightly but that isn't necessary.